### PR TITLE
[BUG] Move prediction tensors to CPU between batches in PredictCallback

### LINKED ISSUE #2224: [BUG] Prediction tensors not moved to CPU between batches, causing VRAM accumulation
**Describe the bug**

When running `model.predict()` on a CUDA device, `PredictCallback.on_predict_batch_end` collects each batch output into self.predictions without calling `.detach()` or `.cpu()`:

[Bug Lives Here:](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/callbacks/predict.py#L69)

```python
self.predictions.append(processed_output)
```

Every prediction tensor stays pinned in VRAM until `on_predict_epoch_end` runs at the very end of inference. The same issue affects `return_info` tensors at [here](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/callbacks/predict.py#L73-L79).

While reviewing I found that, PredictCallback was introduced in [#1984](https://github.com/sktime/pytorch-forecasting/pull/1984) as the v2 predict path and the v1 base model had an explicit guard at [here](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/models/base/_base_model.py#L310-L311) that moved outputs off-device after every batch:

```python
out = move_to_device(detach(out), "cpu")
x   = move_to_device(detach(x),   "cpu")
```

But this pattern was not carried over into the v2 callback.

&nbsp;

**To Reproduce**

Full repro script (any CUDA GPU): [Gist here](https://gist.github.com/StrikerEureka34/19f78514c1a3fbb39b7f353d67599f37)
Result I got while running it locally are:-

```python
# device: cuda
# VRAM before loop: 0.0 MB
# batches collected : 127
# tensor shape      : torch.Size([64, 24])
# tensor device     : cuda:0
# VRAM held by 127 prediction tensors: 9.44 MB
# Expected: ~0 MB (tensors should be on CPU). Got: 9.44 MB on GPU.
# With a real FM dataset (1000+ series, multivariate), this accumulates to OOM.
```

&ensp;

**Expected behavior**

So here `cb.predictions[0].device` should be cpu after each batch. But right now it is `cuda:0`, which means every batch tensor accumulates in VRAM for the entire inference run.

&ensp;

**Proposed Fix**

Add '.detach().cpu()' before each append at [here](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/callbacks/predict.py#L69) for predictions, and at [here](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/callbacks/predict.py#L73-L79) for return_info tensors (y, decoder_lengths) and a dict comprehension over x since it contains mixed tensor and non-tensor values.

&ensp;

**Additional context**

For FM-scale workloads(#1959), inference over thousands of series runs hundreds of batches per call. 
This accumulation is linear, for example 127 batches on this test already held 9.44 MB above model weight allocations. 

What is worse is that on a smaller GPU or with multivariate targets this quietly exhausts VRAM mid-inference with no message pointing to PredictCallback as the cause, which will cause a lot of confusion as it did to me while working with a bigger timeseries data.

If this bug is verified, from the reproducible script as attached, I'm happy to submit a PR for the same, thanks!
@phoeenniixx @fnhirwa @PranavBhatP

### PR DESCRIPTION
**Pull request by StrikerEureka34**
#### Reference Issues/PRs
Fixes #2224

#### What does this implement/fix? Explain your changes.
As described in #2224, tensors collected in `on_predict_batch_end` were staying on the GPU between batches causing VRAM to grow linearly across the inference run.

Fixed by applying `move_to_device(detach(...), "cpu")` at each append site using
utilities already in pytorch_forecasting.utils, the same pattern already used in the v1 base model.  `x_cpu` is computed once and reused for both x and decoder_lengths to avoid reading from the GPU dict twice.
&nbsp;
#### What should a reviewer concentrate their feedback on?
Just the changes in `on_predict_batch_end` in `pytorch_forecasting/callbacks/predict.py`.
Nothing else was touched.
&nbsp;
#### Did you add any tests for the change?
Yes, I've added `tests/test_callbacks.py` with 4 tests covering prediction mode, raw mode dict output, the x dict, and y/decoder_lengths.

There was no existing test file for callbacks so created a standalone `test_callbacks.py` directly under `tests/`, following the same pattern as `test_metrics.py` rather than adding a subdirectory for a single file.

Since these run on CPU (no CUDA needed in CI), the device check alone would pass trivially even without the fix. 
To get around this, I've created tensors as non-leaf (`requires_grad=True` + an addition op) so they have a real `grad_fn`. 
The `assert grad_fn is None` then only passes if `.detach()` actually ran, making these tests meaningful on CPU too.
&nbsp;
#### PR checklist
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG].
- [x] Added/modified tests
- [x] Used pre-commit hooks when committing

### CODE REVIEW

**Comment by StrikerEureka34** [comment]
Hi @fnhirwa @phoeenniixx , I think this should be a complete fix, it uses the existing move_to_device and detach utilities from pytorch_forecasting.utils and includes tests to avoid silent regression.

I reached out to @FarahAbdo to incorporate anything useful from #2226 but haven't heard back yet, ready to do so if they respond.

Would really appreciate your review when you get a chance, thanks!

File: pytorch_forecasting/callbacks/predict.py [hunk 1]
```diff
@@ -66,17 +68,20 @@ def on_predict_batch_end(
...
+        # detach(x) copies the dict; move_to_device won't mutate the original batch
+        x_cpu = move_to_device(detach(x), "cpu")
 
         for key in self.return_info:
             if key == "x":
-                self.info[key].append(x)
+                self.info[key].append(x_cpu)
             elif key == "y":
-                self.info[key].append(y[0])
+                self.info[key].append(move_to_device(detach(y[0]), "cpu"))
```
Comment (fnhirwa) [inline pytorch_forecasting/callbacks/predict.py]: Just a simple modification, can we try to isolate the movement to the device from the append function itself?

just something like 
```python
y_cpu = move_to_device(detach(y[0]), "cpu")
# then append
self.info[key].append(y_cpu)
```

Comment (StrikerEureka34) [reply pytorch_forecasting/callbacks/predict.py]: > Just a simple modification, can we try to isolate the movement to the device from the append function itself?

Thankyou so much for the review!
Makes sense, will do the same for y_cpu and index_cpu. Also planning to guard x_cpu to only compute when actually needed.

**Comment by StrikerEureka34** [comment]
I've made the changes as requested, please let me know if this implementation holds, thanks! @fnhirwa @phoeenniixx

**Review by fnhirwa — APPROVED** [review]
looks good to me
